Fix two mod_aggregate bugs in pkg states

terminalmage opened the pull request:
1. Don't aggregate both ``name``/``pkgs`` and ``sources`` with one another in pkg states

   Doing so will cause problems trying to install as Salt does not support installing from both binary packages and repository packages in the same call to ``pkg.install``.

   The fix here is to aggregate ``name``/``pkgs`` and ``sources`` separately. Resolves #40219.

2. Properly aggregate ``version`` when passed with ``name``

   This fixes an issue in which states which contained a single package name (no ``pkgs`` or ``sources``) paired with a ``version`` argument, were having the version ignored. This is because just the name was being added to the aggregated ``pkgs``.

   Since we're munging low data to aggregate states, this bug wouldn't even result in the state failing if a different version were installed.

terminalmage committed:
Don't aggregate both name/pkgs and sources in pkg states

Doing so will cause problems trying to install as Salt does not support
installing from both binary packages and repository packages in the same
call to pkg.install.

terminalmage committed:
Properly aggregate version when passed with name

This fixes an issue in which states which contained a single package
name (no "pkgs" or "sources") paired with a "version" argument, were
having the version ignored. This is because just the name was being
added to the aggregated "pkgs".

Since we're munging low data to aggregate states, this bug wouldn't even
result in the state failing if a different version were installed.

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @thatch45, @twangboy and @garethgreenaway to be potential reviewers.

terminalmage committed:
Pop off the version when aggregating pkg states

This squelches a spurious warning that we get due to rewriting the low
data such that the name and version are aggregated under the "pkgs" arg,
with the "version" key remaining in the low chunk.

terminalmage commented:
These can be verified using Docker container **terminalmage/issues:40219**. Run the following command:

```
docker run --rm -v ~/path/to/git/checkout:/testing terminalmage/issues:40219 salt-call state.apply test
```

Replace ``~/path/to/git/checkout`` with the path to a git checkout of Salt. Doing so [with this pull request checked out locally](https://help.github.com/articles/checking-out-pull-requests-locally/) will result in the version of zsh configured in the SLS (the older version, ``5.0.2-25.el7``) being installed, while doing so without this fix in place would install the newer version from CentOS' ``updates`` repo, version ``5.0.2-25.el7_3.1``.

```
% docker run --rm -v ~/git/salt/main:/testing terminalmage/issues:40219 salt-call pkg.list_repo_pkgs zsh
local:
    ----------
    base:
        ----------
        zsh:
            - 5.0.2-25.el7
    updates:
        ----------
        zsh:
            - 5.0.2-25.el7_3.1
```

gtmanfred commented:
Looks good! Thanks!
